### PR TITLE
Adds has_dataset method to Client

### PR DESCRIPTION
**Pull request by jakirkham**
Adds the `has_dataset` method to `Client`.

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -1826,6 +1826,18 @@ def list_datasets(self, **kwargs):
...
+        """
+        Checks for named dataset available on the scheduler
+
+        See Also
+        --------
+        Client.publish_dataset
+        Client.get_dataset
+        Client.list_datasets
+        """
+        return name in self.list_datasets()
```
Comment (mrocklin) [inline distributed/client.py]: Perhaps we should just leave this to the user to do on their own?

Comment (jakirkham) [reply distributed/client.py]: Adjusted this to push the check into `PublishExtension`.

Though would be better great if `get_dataset` behaved a bit more like `get` for a `Mapping` instance by taking a `default` value. ATM checking for a dataset that doesn't exist even if one uses a `try...except` triggers a lengthy amount of logging about the `name` in question not being present.

**Comment by jakirkham** [comment]
Any thoughts on the updated version?

**Comment by mrocklin** [comment]
Everything here things consistent to me.  I think my main question here is where do we draw the line between functions that we implement and add to Client/Scheduler and functions that we leave to the user.  There is some cost to keeping around several methods on the Client.  I would not want to implement all of the MutableMapping API on the Client for published datasets, for example.

As an alternative, I wonder if we might separate out the publish API to a separate `MutableMapping` interface.  Someone suggested this previously (maybe you?).  I would be much less conservative here if this API was not strongly attached to the Client.

**Comment by martindurant** [comment]
Don't we already have this as `'mydataname' in client.datasets`? I agree that `client.datasets.has` would be less intrusive than `client.has_dataset`. In fact, I suspect that the publish functions, as opposed to the `dataset` attribute access, have been kept around only for backward compatibility.

**Comment by mrocklin** [comment]
Ah ha, yes, that thing.

**Comment by jakirkham** [comment]
An alternative would be PR ( https://github.com/dask/distributed/pull/1887 ), which might be better for my use case. However it seems to not quite exhibit the behavior desired yet. Also that PR has a better description of the motivation.

**Comment by jakirkham** [comment]
True. Was looking at the `MutableMapping` option as well. Does look nicer for my use case. Only realized that was an option today.

Think the main issue left is the `get_dataset` trace. PR ( https://github.com/dask/distributed/pull/1888 ) should improve on that (if we are ok with that).

Given all of this, will close this PR out. Thanks everyone. :)

**Pull request closed without merging**